mini-projects/01-device-driver/pico-image/src/main.rs

```diff
@@ -91,203 +148,264 @@ fn main() -> ! {
-                (rsg_led_states.0 == PinState::High) as u8,
-                (rsg_led_states.1 == PinState::High) as u8,
-                (rsg_led_states.2 == PinState::High) as u8
-            )
-            .expect("String buffer has been conservatively allocated to fit full payload");
-
-            // Only possible error is when USB Buffer is full, which just means
-            // that this specific message will be dropped.
-            let _ = serial.write(debug_text.as_bytes());
+        if gpios.top.interrupt_status(EdgeLow) {
+            if current_time - *DEBOUNCE_TOP > debounce_time {
+                pos_diff.0 -= 1;
+                pos_diff.1 -= 1;
+                *DEBOUNCE_TOP = current_time;
+            }
+            gpios.top.clear_interrupt(EdgeLow);
+        }
+        if gpios.right.interrupt_status(EdgeLow) {
+            if current_time - *DEBOUNCE_RIGHT > debounce_time {
+                pos_diff.0 += 1;
+                pos_diff.1 += 1;
+                *DEBOUNCE_RIGHT = current_time;
+            }
+            gpios.right.clear_interrupt(EdgeLow);
+        }
+        if gpios.bottom.interrupt_status(EdgeLow) {
+            if current_time - *DEBOUNCE_BOTTOM > debounce_time {
+                pos_diff.0 -= 1;
+                pos_diff.1 += 1;
+                *DEBOUNCE_BOTTOM = current_time;
+            }
+            gpios.bottom.clear_interrupt(EdgeLow);
         }
 
-        // Check for new data
-        if usb_dev.poll(&mut [&mut serial]) {
-            let mut buf = [0u8; 64];
-            match serial.read(&mut buf) {
-                Err(_e) => {
-                    // Do nothing
-                }
-                Ok(0) => {
-                    // Do nothing
-                }
-                Ok(count) => {
-                    let serial_cmd = match core::str::from_utf8(&buf[..count]) {
-                        Ok(s) => s.trim(), // If valid UTF-8, convert to String
-                        Err(_) => {
-                            "" // Returning an empty string for invalid UTF-8
-                        }
-                    };
-
-                    if serial_cmd.contains("enable debug") {
-                        in_debug_mode = true;
-                    }
-                    if serial_cmd.contains("disable debug") {
-                        in_debug_mode = false;
-                    }
+        // Update and send positions if DeviceState::Running
+        if let DeviceState::Running = state {
+            player_position.0 += pos_diff.0;
+            player_position.1 += pos_diff.1;
+            let mut message: String<20> = String::new();
+            writeln!(message, "{:?},{:?}", player_position.0, player_position.1).unwrap();
+            let _ = usb_container
+                .as_mut()
+                .unwrap()
+                .serial
+                .write(message.as_bytes());
+            let _ = usb_container.as_mut().unwrap().serial.flush();
+        }
+    });
```

Comment:
in general, it's best practice to keep interrupts as low compute as possible just to minimize the chance of interrupts stacking on top of each other, I probably would have defaulted to keeping the serial commands in the main loop of the code, and just updating the global states of each button within the interrupt